blob.go: adapt code to use helpers for paths

jf opened the pull request.

ahmetalpbalkan commented:
I actually have a better plan. Do you mind creating methods like `pathForContainer` and `pathForBlob` then we can get rid of all those duplication of fmt.Sprintfs and have the path logic only in one place.

jf commented:
Well now that *is* better alright!

jf commented:
@ahmetalpbalkan are you ok with these methods being called, and the results used directly without any intermediate "`path`" variable? There are no error returns with fmt.Sprintf; it always returns a string.

so eg.
```
uri := b.client.getEndpoint(blobServiceName, b.pathForBlob(container, name), url.Values{})
```

instead of
```
path := fmt.Sprintf("%s/%s", container, name)
uri := b.client.getEndpoint(blobServiceName, path, url.Values{})
```

ahmetalpbalkan commented:
yeah, these can go to blob.go as global non-exported functions. there's no need to make it functions of BlobStorageClient.

jf commented:
Actually we dont need a `pathForContainer`, do we? Just `pathForBlob`.

ahmetalpbalkan commented:
you actually do. what happens when tomorrow we want to address containers as `/{container}` in the code? :) that logic must be only in one place and it should be testable. that's why these helpers are crucial.

jf commented:
eh... well I guess you're right. I was focussed on the present. :}

jf commented:
@ahmetalpbalkan , this is what I have so far (minus the tests). Let me know what you think.

Note: I've added a `/` at the front of the constructed paths because that's what will happen anyway (see http://golang.org/src/net/url/url.go?s=12230:12259#L452, line 481). I'm shortcutting the code, and making things clearer imo.

ahmetalpbalkan commented:
Nope please don't add that. That duplicates the job of net/url package and might have introduced bugs elsewhere. I wouldn't do that unless it's very necessary.  (Did you also run the test suite with this change?)

jf commented:
Yes I did (added that in my comment after I first posted it). I don't think it really duplicates the work of the code in net/url, though. I see that line as just a precaution, or "helper" of sorts ("if the slash isn't prepended, we'll prepend it").

ahmetalpbalkan commented:
Okay I guess it's okay unless it makes requests with double slashes (//) or generates GetBlobURL or SAS URL like that. These are already covered by the tests. I will take a final look and then merge. Thanks @jf.

ahmetalpbalkan commented:
Oh and I realized pathForBlob can reuse pathForContainer. :smile: also having simple unit tests would be really great. Thanks.

jf commented:
ha, I never really considered that! Do you want to keep them separate, though? or are we worrying too much?

I'll get some tests in.

jf commented:
ok I've just added the tests. Had to figure out (took some troubleshooting!) why no matter what I did, the tests I added *weren't* running. So it turns out I found a bug in blob_test.go, and fixed that too. I'm choosing to keep the fix in a separate commit, because I think it's important enough to point out. Prior to this, we weren't running the test for blobSASStringToSign() (and it was wrong to boot, too), apparently!

ahmetalpbalkan commented on clients/storage/blob_test.go:
```diff
@@ -20,7 +20,21 @@ import (
+func Test_pathForBlob(t *testing.T) {
+	out := pathForBlob("foo", "blob")
+	if expected := "/foo/blob"; out != expected {
+		t.Errorf("Wrong pathForContainer. Expected: '%s', got: '%s'", expected, out)
```
pathForBlob

jf replied:
:}

jeffmendoza commented:
@jf can you sign a CLA at https://cla.msopentech.com/ ?

ahmetalpbalkan commented:
Please rebase to include the CI commit, otherwise test will fail.

jf commented:
@jeffmendoza is there a more specific privacy policy for developers/open source contributors? I don't see how getting my address, for example, is necessary

jeffmendoza commented:
Looks like there is a link at the end of the page to here: https://msopentech.com/privacy-policy/

jf commented:
I'm aware and have already looked through it. That's a generic one for all of "MS Open Tech" - "Msopentech.com and other Microsoft Open Technologies, Inc. (“MS Open Tech”) websites and services of MS Open Tech that collect data and display these terms, as well as its offline product support services.". Perhaps some of these will have a legitimate reason for asking for an address.... but as a "pure github contributor and nothing else", I see no reason for Microsoft to ask for my address.

jf commented:
Anyway, I've filled in the form. But haven't received any email yet. Is it automated? or manual? (so ie. a delay is normal)

ahmetalpbalkan commented:
@hopetobelievein is CLA approved for @jf ?

jeffmendoza commented:
@jf Looks like our new system just took over, can you try: https://cla2.msopentech.com/

jf commented:
Ok I'm going to give it one more try. But if this doesn't go through, I'm out for submitting PRs.

ahmetalpbalkan commented:
@jeffmendoza is CLA approved for @jf?

jf commented:
I am approved already, @ahmetalpbalkan . But dunno if this is good enough for you, or if you're required to hear from one of the folks you've asked first (in which case, we need to wait on them).

ahmetalpbalkan commented:
@jf ah okay, it is just I don't have access to the interface that shows if you're approved or not. I did not realize you got the approval email from the context above.

jeffmendoza commented:
@jf Thanks for bearing with us
